nucypher/blockchain/eth/trackers/dkg.py

```diff
@@ -62,47 +72,64 @@ def __init__(
             operator=operator,
             web3=operator.coordinator_agent.blockchain.w3,
             contract=contract,
-            events=events,
+            events=all_events,
             actions=actions,
             persistent=persistent,
             timeout=self.coordinator_agent.get_dkg_timeout(),
         )
 
     def _get_identifier(self, event: AttributeDict) -> str:
-        return event.args.ritualId
+        event_type = getattr(self.contract.events, event.event)
+        if event_type in self.handover_event_types:
+            return f"{self.HandoverParticipationState.PREFIX}{event.args.ritualId}"
+        else:
+            return str(event.args.ritualId)
 
     def _action_required_based_on_participation_state(
-        self, participation_state: DkgParticipationState, event: AttributeDict
+        self,
+        participation_state: Union[DkgParticipationState, HandoverParticipationState],
+        event: AttributeDict,
     ) -> bool:
         # Let's handle separately handover events and non-handover events
         event_type = getattr(self.contract.events, event.event)
         if event_type in self.handover_event_types:
             # handover modifies existing ritual metadata; so we need to proactively prune it
-            # during handover process and at the end to avoid having any stale metadata
-            # in the cache
+            # during handover process for ALL associated nodes
+            # (part of handover, ALL existing participants in ritual) to avoid having
+            # any stale metadata in the metadata storage cache
             self.operator.prune_ritual_metadata_due_to_handover(
                 event.args.ritualId
             )
 
-            if event_type == self.contract.events.HandoverFinalized:
-                # pruning metadata is sufficient when Handover is finalized;
+            if event_type in [
+                self.contract.events.HandoverFinalized,
+                self.contract.events.HandoverCanceled,
+            ]:
+                if (
+                    event_type == self.contract.events.HandoverCanceled
+                    and event.args.incomingParticipant == self.operator.checksum_address
+                ) or (
+                    event_type == self.contract.events.HandoverFinalized
+                    and event.args.departingParticipant
+                    == self.operator.checksum_address
+                ):
+                    # if handover is canceled/finalized we need to reset the participation state
+                    participation_state.participating = False
+
+                # pruning metadata/cleanup is sufficient when Handover is finalized or canceled;
                 # no further action required
                 return False
```

Comment:
Pretty sure I'm misunderstanding the handover logic here, but is it actually possible to have a HandoverCanceled or HandoverFinalized but with the node still participating of it?

In other words: is this the same that just


```suggestion
            if event_type in [
                self.contract.events.HandoverFinalized,
                self.contract.events.HandoverCanceled,
            ]:
                # pruning metadata/cleanup is sufficient when Handover is finalized or canceled;
                # no further action required
                return False
```

Reply:
Participation pertains to the signing cohort not the handover.

> is it actually possible to have a HandoverCanceled or HandoverFinalized but with the node still participating of it?

Yes. This is the case for nodes that are in the cohort and not part of handover (neither departing or incoming participant). 

The logic behind the checks are
1. If handover was canceled then:
- if the node was the incoming participant, then it is still not participating in the 
- if the node was the departing participant, then it is still participating in the cohort (stays the same)
- if the node was neither then its participation remains the same (no change)
2. If handover was finalized then:
- if the node was the departing participant, then it is no longer participating in that cohort
- if the node was the incoming participant, then it is now participating in that cohort (stays the same)
- if the node was neither then its participation remains the same (no change)

So the first bullet point for 1) and 2) is what this extra check in the code is performing.

Reply:
oooh, got it. Yes, as you are pointing, I thought participant meant that you are a participant of the handover instead of the ritual. Now it's clear, thanks.

Reply:
I'll update some of the class documentation to help clarify that. Based on the naming I can see where the confusion came from.

Reply:
@manumonti, added https://github.com/nucypher/nucypher/pull/3638/commits/550dcda3909e2576379a6c8c51e2b62d1c5f3919.